### ISSUE TITLE
Can't train model with dialogue length limit=500 on even NVIDIA Volta GPU

### ISSUE THREAD

**Issue by dimeldo** [chin-gyou/improving-variational-encoder-decoders-in-dialogue-generation#1]
Hi,

I want to train my dataset with dataproducer.py limit set to 500. It seems like https://github.com/snakeztc/NeuralDialog-CVAE can easily handle that and even bigger (limit=1000) with batch size 128. I'm using NVIDIA Volta GPU.

There'es some inefficiency going on here that I can't seem to solve.
Do you have any idea how can I go about solving this?

Thanks.

**Comment by dimeldo** [comment]
Maybe it has something to do with this warning:

`UserWarning: Converting sparse IndexedSlices to a dense Tensor of unknown shape. This may consume a large amount of memory.
  "Converting sparse IndexedSlices to a dense Tensor of unknown shape. "
`

From [StackOverflow](https://stackoverflow.com/questions/35892412/tensorflow-dense-gradient-explanation):

> This warning is printed when a sparse tf.IndexedSlices object is implicitly converted to a dense tf.Tensor. This typically happens when one op (usually tf.gather()) backpropagates a sparse gradient, but the op that receives it does not have a specialized gradient function that can handle sparse gradients. As a result, TensorFlow automatically densifies the tf.IndexedSlices, which can have a devastating effect on performance if the tensor is large.

> To fix this problem, you should try to ensure that the params input to tf.gather() (or the params inputs to tf.nn.embedding_lookup()) is a tf.Variable. Variables can receive the sparse updates directly, so no conversion is needed. Although tf.gather() (and tf.nn.embedding_lookup()) accept arbitrary tensors as inputs, this may lead to a more complicated backpropagation graph, resulting in implicit conversion.
